client/network/src/service.rs

```diff
@@ -205,12 +217,55 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkWorker<B, H> {
 			]
 		};
 
+		// Initialize the permissioned nodes
+		let mut init_allowlist = None;
+		if params.permissioned_network {
+			let id = BlockId::hash(params.chain.info().best_hash);
+			let raw_allowlist = match params.chain.storage(&id, &StorageKey(well_known_keys::NODE_ALLOWLIST.to_vec())) {
+				Ok(Some(r)) => r,
+				Ok(None) => { 
+					info!(
+						target: "sub-libp2p",
+						"🏷  err in get init storage -------------"
+					);
+					StorageData(vec![])
+				},
+				Err(_) => return Err(Error::InvalidStorage), // TODO log error
+			};
+			let node_allowlist: Vec<NodePublic> = match Decode::decode(&mut &raw_allowlist.0[..]) {
+				Ok(r) => r,
+				Err(_) => {
+					info!(
+						target: "sub-libp2p",
+						"🏷  err in init decode------------- {:?}",
+						raw_allowlist
+					);
+					vec![]
+				}, // TODO log error
+			};
+			// Transform to PeerId
+			let peer_ids = node_allowlist.iter()
+				.filter_map(|pubkey| {
+					info!(
+						target: "sub-libp2p",
+						"🏷  init deocoding node id now: {:?}, {:?}",
+						Ed25519PublicKey::decode(&pubkey.0),
+						&pubkey.0
+					);
+					Ed25519PublicKey::decode(&pubkey.0).ok()
+				})
+				.map(|pubkey| PublicKey::Ed25519(pubkey).into_peer_id())
+				.collect();
+			init_allowlist = Some(peer_ids);
+		}
```

Comment:
I believe that this entire block is extremely similar to the one that you have added to `sc-service` as well, right?
If so, I'd prefer if we could initialize the network to reject everything by default, then let `sc-service` configure which peer we're allowed to connect to.

Reply:
The import block stream is triggered by new block, if reject everything in network initialization, there will be no further block import stream. And if accept everything in network initialization, then every time the node is restart, all the peers are free to be connected.

Reply:
Is is make sense if I put the initialization logic at the start of the function [build_network_future](https://github.com/paritytech/substrate/blob/ed4f7a113073c08cba9020c3f0dd6c8f0441f857/client/service/src/lib.rs#L193)?

Reply:
You can put that code at the beginning of [this `loop`](https://github.com/paritytech/substrate/blob/6d6944435f816777b33a5ee274be845c2ae24065/client/service/src/lib.rs#L225) inside of an `if refresh_allow_list {`.  `refresh_allow_list` would be true at initialization, would be set to false when we update the list, and would be set to true on each new block.